### PR TITLE
`ultralytics 8.3.230` Fix color conversion for Tensor inputs and PIL plots

### LINKED ISSUE #22732: Prediction result plot colors
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When [plotting a prediction result](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot), the colors of the original image does not match if the input image was fed to the model as a PIL.Image object. However, if the input image was fed as a tensor, then the plot colors are correct.

In the example below, I run prediction on a completely red image. In `TEST 1`, a feed the image as a tensor, and [`Results.plot`](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot) correctly displays a red image. However, in `TEST 2`, i feed the image as a PIL.Image object, and [`Results.plot`](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot) shows a completely blue image.

### Environment

```
Ultralytics 8.3.204 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4090, 24072MiB)
Setup complete ✅ (32 CPUs, 124.9 GB RAM, 701.0/3665.4 GB disk)

OS                     Linux-6.14.0-27-generic-x86_64-with-glibc2.39
Environment            Linux
Python                 3.11.13
Install                git
Path                   <REDACTED>
RAM                    124.95 GB
Disk                   701.0/3665.4 GB
CPU                    AMD Ryzen 9 7950X3D 16-Core Processor
CPU count              32
GPU                    NVIDIA GeForce RTX 4090, 24072MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.10.1>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.1.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.15.2>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.0.0
polars                 ✅ 1.34.0
ultralytics-thop       ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

```py
from IPython.display import display
import torchvision.transforms.functional as tr
from ultralytics import YOLO
import torch
from PIL import Image

model = YOLO("yolo11n.pt")
image_array = torch.stack(
    [
        torch.ones((256, 256)),
        torch.zeros((256, 256)),
        torch.zeros((256, 256)),
    ]
)
image = tr.to_pil_image(image_array)
display(image)  # all red

# TEST 1 ========================================
r = model(image_array.unsqueeze(0), verbose=False)[0]
display(tr.to_pil_image(r.plot()))  # all red
display(Image.fromarray(r.plot()))  # all red
display(r.plot(pil=True))  # all red

# TEST 2 ========================================
r = model(image, verbose=False)[0]
display(tr.to_pil_image(r.plot()))  # all blue
display(Image.fromarray(r.plot()))  # all blue
display(r.plot(pil=True))  # all blue
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
1. Tensor input needs to be converted from RGB to BGR.
2. Plotting with `pil=True` needs to convert from BGR to RGB

Fixes #22732 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes image color handling and plotting outputs across multiple models, ensuring consistent, correct-color visualizations and a cleaner `Results.plot()` API. 🎨✅

---

### 📊 Key Changes
- 🔼 Bumped package version from `8.3.229` to `8.3.230`.
- 🖼️ Updated `Results.plot()` to always use `annotator.result(pil)` so the `pil` flag properly controls whether a NumPy array or PIL image is returned.
- 🎨 Adjusted `Annotator`:
  - Keeps internal images in BGR when using PIL (to match the existing color palette).
  - Changed `result()` to `result(pil=False)`:
    - Returns a BGR NumPy array by default.
    - Returns an RGB PIL image when `pil=True`.
- 🌈 Fixed original-image color ordering in postprocessing for:
  - RT-DETR (`rtdetr/predict.py`)
  - SAM (`sam/predict.py`)
  - YOLO classification (`yolo/classify/predict.py`)
  - YOLO detection (`yolo/detect/predict.py`)
  - When `orig_imgs` is a tensor, it’s converted then channel-flipped (`[..., ::-1]`) to correct BGR/RGB order.

---

### 🎯 Purpose & Impact
- ✅ **Correct colors in visualizations**: Fixes BGR/RGB mismatches so plotted outputs (detections, masks, classifications) look natural and match the original image colors.
- 🧩 **Consistent API behavior**: `Results.plot(pil=True/False)` now reliably returns the type users expect (PIL image vs NumPy array), reducing confusion and bugs in downstream code.
- 📸 **Better compatibility for tensor inputs**: When users pass tensors directly, postprocessing now uses correctly ordered images, improving overlays and saved results across RT-DETR, SAM, and YOLO models.
- 🧪 **More predictable plotting pipeline**: Internal BGR handling is explicit, making future debugging and extension of plotting logic more robust.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review helper, and a Ultralytics engineer will be by shortly to assist further.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing `I have read the CLA Document and I sign the CLA` in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/yolo/detect/predict.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def postprocess(self, preds, img, orig_imgs, **kwargs):
         )
 
         if not isinstance(orig_imgs, list):  # input images are a torch.Tensor, not a list
-            orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)
+            orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)[..., ::-1]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/predict.py:68]: ⚠️ **HIGH**: `convert_torch2numpy_batch` already outputs RGB BHWC arrays, so appending `[..., ::-1]` converts tensors back to BGR while list-based inputs remain RGB. This reintroduces the color swap you’re trying to fix and makes tensor vs. list paths inconsistent across the detect, classify, RT-DETR, and SAM predictors. Please keep the tensors in RGB and apply any necessary conversion at a single, consistent point in the pipeline (same fix needed at R58 in `rtdetr/predict.py`, R505 in `sam/predict.py`, and R84 in `yolo/classify/predict.py`).

**Suggested change:**
```suggestion
            orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)
```

File: ultralytics/utils/plotting.py [hunk 2]
```diff
@@ -207,7 +207,7 @@ def __init__(
             elif im.shape[2] > 3:  # multispectral
                 im = np.ascontiguousarray(im[..., :3])
         if self.pil:  # use PIL
-            self.im = im if input_is_pil else Image.fromarray(im)
+            self.im = im if input_is_pil else Image.fromarray(im[..., ::-1])
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/plotting.py]: ⚠️ **HIGH**: `Image.fromarray` already expects RGB channel order. Unconditionally reversing channels here will turn valid RGB images into BGR and skew every PIL-based plot while also double-flipping tensors processed earlier. If some inputs still arrive in BGR, convert them before they reach the annotator instead of flipping everything here.

**Suggested change:**
```suggestion
            self.im = im if input_is_pil else Image.fromarray(im)
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The new channel-reversal logic now flips already-correct RGB images back to BGR in every tensor-based predictor and in the PIL plotting path, breaking colors for the majority of use cases. Keep `convert_torch2numpy_batch` outputs in RGB and only normalize channel order once instead of double-flipping downstream.

💬 Posted 2 inline comments

File: ultralytics/models/yolo/detect/predict.py (hunk 1 above)
Comment (Y-T-G) [reply ultralytics/models/yolo/detect/predict.py:68]: It's supposed to be BGR just like numpy arrays

File: ultralytics/utils/plotting.py (hunk 2 above)
Comment (Y-T-G) [reply ultralytics/utils/plotting.py]: `im` is a BGR numpy array. So it needs to be converted to RGB.

**Comment by glenn-jocher** [comment]
@Y-T-G is this aligned with https://docs.ultralytics.com/modes/predict/#inference-sources? The Docs say that any Tensor images need to be passed in RGB order.

Screenshot 2025-11-20 at 1 35 54 PM

**Comment by Y-T-G** [comment]
@glenn-jocher Yes, the Tensor input is RGB. Sinc the Tensor input is RGB, it needs to be converted to a BGR numpy array when creating `results`. But currently, it keeps it in original channel order, so the output has the wrong channel order.

```python
from ultralytics import YOLO, ASSETS
import cv2

model = YOLO("yolo11n.pt")
results1 = model(ASSETS / "bus.jpg")

img = cv2.imread(str(ASSETS / "bus.jpg"))
tensor = model.predictor.preprocess([img])  # RGB tensor
results2 = model(tensor)

results2[0].show()
```

image

**Comment by UltralyticsAssistant** [comment]
What an awesome PR to get merged – huge thanks @Y-T-G for driving this, and @glenn-jocher for the collaboration behind the scenes!  

> “Details make perfection, and perfection is not a detail.” – Leonardo da Vinci  

This work on consistent color handling and a cleaner `Results.plot()` API is exactly that kind of detail: subtle on the surface, but it dramatically improves how users experience RT-DETR, SAM, and YOLO visualizations across the board. Correct colors, predictable outputs, and a more robust plotting pipeline will save a lot of people time and confusion.  

Really appreciate the care and precision you both brought to this – contributions like this keep the Ultralytics ecosystem polished and a joy to use. 🙌
